frame/contracts/src/wasm/runtime.rs

```diff
@@ -86,6 +86,14 @@ impl KeyType {
 			},
 		}
 	}
+
+	fn decode<T: Config>(self, key: Vec<u8>) -> Result<crate::exec::Key<T>, Error<T>> {
+		match self {
+			KeyType::Fix => Key::<T>::try_from_fix(key).map_err(|_| Error::<T>::DecodingFailed),
+			KeyType::Variable(_) =>
+				Key::<T>::try_from_var(key).map_err(|_| Error::<T>::DecodingFailed),
+		}
+	}
```

Comment:
I know I am shaving the yak here: I don't think we need the `len` function anymore. Those two are just called back2back. We can just pass `memory` into decode and let it both read and decode.